Using Vault PKI to generate a cluster CA (#41)
strimzi/strimzi.github.io: _posts/2019-11-09-using-vault-with-strimzi.md

It is this last point, certificates, which we are going to be looking at. Specifically, we will be using the [Vault PKI Secrets Engine](https://www.vaultproject.io/docs/secrets/pki/index.html) to create a CA, which we will then install instead of using CA certificates and private keys generated by the [Strimzi Cluster Operator](http://strimzi.io/docs/master/#cluster-operator-str).

# SSL, Strimzi & Security

## The `incubator` Kafka charts

Like me, you may have tried configuring SSL for Kafka using the `incubator` Helm charts (you can find them [here](https://github.com/helm/charts/tree/master/incubator/kafka)), and realised that in fact this is _far from trivial_. I had initially configured an `initContainer` to get certificates from the Vault PKI backend using the Vault CLI, which did indeed work. However, in addition to the startup scripts in the main broker container and another custom `initContainer` I had configured to set up rack awareness, this caused the the pod startup times to become quite long!

> At this point I had configured the brokers and clients with SSL, but what about the Zookeepers? Zookeeper does not support TLS, which is problematic if, for example, you are a financial organsiation streaming sensitive data through your Kafka cluster. Even if you're not a fintech company, but you're a sensible individual with security concerns, this might be a deal breaker for you!

Reviewer: ```suggestion
At this point I had configured the brokers and clients with SSL, but what about the Zookeepers? Zookeeper does not support TLS, which is problematic if, for example, you are a financial organisation streaming sensitive data through your Kafka cluster. Even if you're not a fintech company, but you're a sensible individual with security concerns, this might be a deal breaker for you!
```

> The Cluster Operator automatically sets up TLS certificates to enable encryption and authentication within your cluster. It also sets up other TLS certificates if you want to enable encryption or TLS authentication between Kafka brokers and clients.

By default, Strimzi generates its own CA and stores the CA certificate and private key as Kubernetes secrets. Using these two secrets, Strimzi generates certificates for all of the components in the cluster, which in turn ensures all communication is encrypted!

## Bringing your own CA

Now, in order to bring your own CA, you need to do this yourself and add the necessary certificate and key as Kubernetes secrets manually, adding the following labels: 

```bash
  strimzi.io/kind="Kafka"
  strimzi.io/cluster="${clusterName}"
```

# Why Vault?

Using Vault to create your own CA arguably gives you more control over your TLS configuration.

> Consider a Kafka cluster with a Kafka client you have developed. Your cluster requires that clients authenticate themselves using SSL, i.e. `ssl.client.auth="required"`.

Reviewer: Be clearer that it's Strimzi which brings the `ssl.client.auth="required"` requirement.

```suggestion
Consider a Strimzi Kafka cluster with a Kafka client you have developed. The cluster requires that clients authenticate themselves using SSL, i.e. `ssl.client.auth="required"`.
```

Vault integrates particularly well with Kubernetes, for example using the [Kubernetes Auth Method](https://www.vaultproject.io/docs/auth/kubernetes.html).

See the Vault documentation for a full list of features:

* [Secrets Engines](https://www.vaultproject.io/docs/secrets/index.html)

* [Auth Methods](https://www.vaultproject.io/docs/auth/index.html)

# Configuration of the Vault PKI secrets engine

Assuming you have a Vault cluster running, you can [create your own CA by doing the following](https://learn.hashicorp.com/vault/secrets-management/sm-pki-engine):

```bash
# Creating the root CA:
# First, enable the pki secrets engine at the pki path:
$ vault secrets enable pki

# Tune the pki secrets engine to issue certificates with a maximum time-to-live (TTL) 

> #   of 87600 hours.

Reviewer: ```suggestion
#   of 87600 hours (10 years).
```

$ vault secrets tune -max-lease-ttl=87600h pki

# Generate the root certificate and save the certificate in CA_cert.crt.
$ vault write -field=certificate pki/root/generate/internal common_name="example.com" \
       ttl=87600h > CA_cert.crt

> # This generates a new self-signed CA certificate and private key. Vault will automatically

Reviewer: > This generates a new self-signed CA certificate and private key.

I assume does the `CA_cert.crt` just contains the certificate, and the key is kept within vault? It might be worth being explicit about that.

Author: @tombentley Yes, precisely. So in this case we never get to see the root's secret key.

#   revoke the generated root at the end of its lease period (TTL); the CA certificate will
#     sign its own Certificate Revocation List (CRL).

# Configure the CA and CRL URLs:
$ vault write pki/config/urls \
       issuing_certificates="http://127.0.0.1:8200/v1/pki/ca" \
       crl_distribution_points="http://127.0.0.1:8200/v1/pki/crl"

# Creating the intermediate CA:
# First, enable the pki secrets engine at the pki_int path:
$ vault secrets enable -path=pki_int pki

# Tune the pki_int secrets engine to issue certificates with a maximum time-to-live (TTL)

> #   of 43800 hours.

Reviewer: ```suggestion
#   of 43800 hours (5 years).
```

  && kubectl label secret -n ${strimziNamespace} ${clusterName}-cluster-ca-cert \
  strimzi.io/kind="Kafka" \
  strimzi.io/cluster="${clusterName}"
```

## Creating your Strimzi cluster

You're almost there! Just a few more steps:

1. Deploy the Strimzi operator
2. Apply your `Kafka` CRD, remembering to set the following parameter:

```yaml
...
spec:
  clusterCa:
    generateCertificateAuthority: false
...
```

### A few things to note...

* The `common_name` cannot contain a wildcard, so set it to something sensible. Otherwise you'll see something like this in your `tls-sidecar`:

```text
LOG5[1:139683950376704]: Service [zookeeper-2181] accepted connection from 127.0.0.1:33576
LOG5[1:139683950376704]: connect_blocking: connected 10.100.241.50:2181
LOG5[1:139683950376704]: Service [zookeeper-2181] connected remote server from 10.100.214.45:37700
LOG4[1:139683950376704]: CERT: Verification error: permitted subtree violation
LOG4[1:139683950376704]: Certificate check failed: depth=0, /O=io.strimzi/CN=zookeeper
LOG3[1:139683950376704]: SSL_connect: 14090086: error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed
```

> * When generating your CA, `type` needs to be `exported`, otherwise the command will _not_ return the private key! [More information on that here.](https://www.vaultproject.io/api/secret/pki/index.html#generate-root)

Reviewer: ```suggestion
* When generating your intermediate CA, `type` needs to be `exported`, otherwise the command will _not_ return the private key! [More information on that here.](https://www.vaultproject.io/api/secret/pki/index.html#generate-root)
```

> * If you would prefer to set up Vault PKI this using the UI or the API, there is documentation for that [here.](https://www.vaultproject.io/docs/secrets/pki/index.html) _Gotcha!_ Remember the importance of the `type`, which needs to be `exported`. The Hashicorp docs use `internal`.

Reviewer: ```suggestion
* If you would prefer to set up Vault PKI this using the UI or the API, there is documentation for that [here](https://www.vaultproject.io/docs/secrets/pki/index.html). _Gotcha!_ Remember the importance of the `type`, which needs to be `exported`. The Hashicorp docs use `internal`.
```

---
layout: post
title:  "Using Vault PKI secrets engine as a CA with the Strimzi Kafka Operator"
date: 2018-11-14
author: chris_vest
---

<!--more-->

# A brief introduction to Vault

If you're here you probably know what [Vault](https://www.vaultproject.io/) is and why it's useful; but for the uninitiated here is a brief description from the creators themselves. The Hashicorp Vault project "secures, stores, and tightly controls access to tokens, passwords, certificates, API keys, and other secrets in modern computing [...]".

So how does Vault fit in with Strimzi and Kafka? Well, Vault has the concept of [`Secrets Engines`](https://www.vaultproject.io/docs/secrets/):

> > Secrets engines are components which store, generate, or encrypt data. Secrets engines are incredibly flexible, so it is easiest to think about them in terms of their function. Secrets engines are provided some set of data, they take some action on that data, and they return a result.\

Reviewer 2: the "\" at the end, is it a typo?

Author: Removed.

> > Some secrets engines simply store and read data - like encrypted Redis/Memcached. Other secrets engines connect to other services and generate dynamic credentials on demand. Other secrets engines provide encryption as a service, totp generation, certificates, and much more.

Reviewer 2: "totp" ? it's a typo I guess ...

Author: It's actually a reference to "Time-based One-time Password algorithm"! I'll add a note.

Reviewer 2: ah!! I didn't get it even because it wasn't quoted or something like this. I think that a note could be better, thanks!

# SSL, Strimzi & Security

## The `incubator` Kafka charts

Like me, you may have tried configuring SSL for Kafka using the `incubator` Helm charts (you can find them [here](https://github.com/helm/charts/tree/master/incubator/kafka)), and realised that in fact this is _far from trivial_. I had initially configured an `initContainer` to get certificates from the Vault PKI backend using the Vault CLI, which did indeed work. However, in addition to the startup scripts in the main broker container and another custom `initContainer` I had configured to set up rack awareness, this caused the the pod startup times to become quite long!

At this point I had configured the brokers and clients with SSL, but what about the Zookeepers? Zookeeper does not support TLS, which is problematic if, for example, you are a financial organisation streaming sensitive data through your Kafka cluster. Even if you're not a fintech company, but you're a sensible individual with security concerns, this might be a deal breaker for you!

## Strimzi & Security

Straight out of the box, [Strimzi provides all of this for you](http://strimzi.io/docs/master/#security-str):

> > Strimzi supports encrypted communication between the Kafka and Strimzi components using the TLS protocol. Communication between Kafka brokers (interbroker communication), between Zookeeper nodes (internodal communication), and between these and the Strimzi operators is always encrypted. Communication between Kafka clients and Kafka brokers is encrypted according to how the cluster is configured. For the Kafka and Strimzi components, TLS certificates are also used for authentication.\

Reviewer 2: the "\" here as well?

Author: It was just for formatting in the markdown preview. Will it cause issues when published?

Reviewer 2: Maybe yes, let's delete this and if needed we'll fix after publication.

Author: Removing it.

> The Cluster Operator automatically sets up TLS certificates to enable encryption and authentication within your cluster. It also sets up other TLS certificates if you want to enable encryption or TLS authentication between Kafka brokers and clients.

By default, Strimzi generates its own CA and stores the CA certificate and private key as Kubernetes secrets. Using these two secrets, Strimzi generates certificates for all of the components in the cluster, which in turn ensures all communication is encrypted!

## Bringing your own CA

Now, in order to bring your own CA, you need to do this yourself and add the necessary certificate and key as Kubernetes secrets manually, adding the following labels: 

```bash
  strimzi.io/kind="Kafka"
  strimzi.io/cluster="${clusterName}"
```

# Why Vault?

Using Vault to create your own CA arguably gives you more control over your TLS configuration.

> Consider a Strimzi Kafka cluster with a Kafka client you have developed. The cluster requires that clients authenticate themselves using SSL, i.e. `ssl.client.auth="required"`.

Reviewer 2: It seems that a Kafka cluster deployed by Strimzi, always required that clients have to authenticate via TLS (please use TLS instead of SSL) but it's not true because you have to configure the `tls` listener in the YAML file for the cluster. I think that you can try to be more clear on this sentence.

LOG5[1:139683950376704]: connect_blocking: connected 10.100.241.50:2181
LOG5[1:139683950376704]: Service [zookeeper-2181] connected remote server from 10.100.214.45:37700
LOG4[1:139683950376704]: CERT: Verification error: permitted subtree violation
LOG4[1:139683950376704]: Certificate check failed: depth=0, /O=io.strimzi/CN=zookeeper
LOG3[1:139683950376704]: SSL_connect: 14090086: error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed
```

* When generating your intermediate CA, `type` needs to be `exported`, otherwise the command will _not_ return the private key! [More information on that here.](https://www.vaultproject.io/api/secret/pki/index.html#generate-root)

* If you would prefer to set up Vault PKI this using the UI or the API, there is documentation for that [here](https://www.vaultproject.io/docs/secrets/pki/index.html). _Gotcha!_ Remember the importance of the `type`, which needs to be `exported`. The Hashicorp docs use `internal`.

* You might be tempted to just use the root CA... But _definitely use an intermediate CA!_ [Why?](https://www.whichssl.com/intermediate-certificates.html)

# Conclusion

> 

Reviewer 2: Why the "Conclusion" section is empty?

Author: Good question! I'll add something here.
